File: backend/voice_agent.py
import asyncio
import base64
import json
import os
import websockets
from deepgram import DeepgramClient, LiveTranscriptionEvents, LiveOptions
# from google import genai will be used. For streaming we can just use litellm or google-genai
import google.generativeai as genai

DEEPGRAM_API_KEY = os.environ.get("DEEPGRAM_API_KEY")
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
ELEVENLABS_API_KEY = os.environ.get("ELEVENLABS_API_KEY")
ELEVENLABS_VOICE_ID = os.environ.get("ELEVENLABS_VOICE_ID", "21m00Tcm4TlvDq8ikWAM")  # Default Rachel

# Configure Gemini
if GEMINI_API_KEY:
    genai.configure(api_key=GEMINI_API_KEY)

class VoiceAgentConnection:
    def __init__(self, twilio_ws, stream_sid, context=""):
        self.twilio_ws = twilio_ws
        self.stream_sid = stream_sid
        self.context = context
        
        self.deepgram = None
        self.elevenlabs_ws = None
        
        self.dg_connection = None
        
        self.transcript_buffer = ""
        self.llm_task = None
        self.tts_task = None
        
        self.is_interrupted = False
        
        # We need a queue to hold TTS chunks before sending them to Elevenlabs
        self.tts_queue = asyncio.Queue()

    async def start(self):
        # Initialize Deepgram
        try:
            deepgram = DeepgramClient(DEEPGRAM_API_KEY)
            self.dg_connection = deepgram.listen.asyncwebsocket.v("1")
            
            self.dg_connection.on(LiveTranscriptionEvents.Transcript, self.on_dg_transcript)
            self.dg_connection.on(LiveTranscriptionEvents.SpeechStarted, self.on_dg_speech_started)
            
            options = LiveOptions(
                model="nova-2",
                language="en-US",
                encoding="mulaw",
                channels=1,
                sample_rate=8000,
                interim_results=False,
                utterance_end_ms="1000",
                vad_events=True
            )
            await self.dg_connection.start(options)
        except Exception as e:
            print(f"Failed to start Deepgram: {e}")

        # Start ElevenLabs WS Connection
        elevenlabs_url = f"wss://api.elevenlabs.io/v1/text-to-speech/{ELEVENLABS_VOICE_ID}/stream-input?model_id=eleven_monolingual_v1&output_format=ulaw_8000"
        try:
            self.elevenlabs_ws = await websockets.connect(elevenlabs_url)
            # send initial config
            initial_msg = {
                "text": " ",
                "voice_settings": {"stability": 0.5, "similarity_boost": 0.8},
                "xi_api_key": ELEVENLABS_API_KEY
            }
            await self.elevenlabs_ws.send(json.dumps(initial_msg))
            asyncio.create_task(self.receive_elevenlabs_audio())
            asyncio.create_task(self.send_tts_queue_to_elevenlabs())
        except Exception as e:
            print(f"Failed to start ElevenLabs WS: {e}")
            
    async def process_audio_from_twilio(self, data):
        """Called when audio is received from Twilio"""
        if self.dg_connection:
            await self.dg_connection.send(data)

    async def on_dg_speech_started(self, *args, **kwargs):
        """User barged in"""
        self.is_interrupted = True
        print("Barge-in detected!")
        # Stop current LLM generation if any
        if self.llm_task and not self.llm_task.done():
            self.llm_task.cancel()
        
        # Clear Twilio audio buffer
        clear_msg = {
            "event": "clear",
            "streamSid": self.stream_sid
        }
        await self.twilio_ws.send_text(json.dumps(clear_msg))
        
        # Reset queue
        while not self.tts_queue.empty():
            try:
                self.tts_queue.get_nowait()
            except asyncio.QueueEmpty:
                break

    async def on_dg_transcript(self, *args, **kwargs):
        """Received transcript from Deepgram"""
        # simplified extraction. In deepgram-sdk it is in kwargs or passed as object
        # The accurate syntax for new SDK:
        result = args[0] if len(args) > 0 else kwargs.get('result')
        if not result: return
        
        sentence = result.channel.alternatives[0].transcript
        if sentence and result.is_final:
            print(f"User heard: {sentence}")
            self.transcript_buffer += " " + sentence
            self.is_interrupted = False
            
            # Start LLM response immediately if enough sentence is gathered.
            # Usually we wait until user completes to speak. We use utterance end or just send.
            if self.llm_task and not self.llm_task.done():
                self.llm_task.cancel()
            self.llm_task = asyncio.create_task(self.generate_response(self.transcript_buffer))
            self.transcript_buffer = ""

    async def generate_response(self, text):
        if not text.strip() or self.is_interrupted: return
        try:
            model = genai.GenerativeModel("gemini-1.5-flash")
            system_instruction = (
                "You are a professional AI phone call representative. "
                "Your objective is defined in the Global Pitch and Client Context below. "
                "Handle the client by following the logic and goals of the pitch. "
                "Do not merely repeat the pitch; use it as a guide for a natural, brief, and professional conversation. "
                "If the client asks something off-topic, politely bring them back to the objective if possible. "
                f"\n\n--- INSTRUCTIONS & CONTEXT ---\n{self.context}\n--- END ---"
            )
            prompt = f"{system_instruction}\n\nUser: {text}\nAI:"
            print(f"Sending to Gemini with prompt length: {len(prompt)}")
            response = await model.generate_content_async(prompt, stream=True)
            
            sentence_chunk = ""
            # chunk tokens on punctuation to send to ElevenLabs
            async for chunk in response:
                if self.is_interrupted: break
                token = chunk.text
                if token:
                    sentence_chunk += token
                    if any(p in sentence_chunk for p in [".", "?", "!", "\n"]):
                        await self.tts_queue.put(sentence_chunk)
                        sentence_chunk = ""
            if sentence_chunk:
                await self.tts_queue.put(sentence_chunk)
        except Exception as e:
            print(f"LLM Error: {e}")

    async def send_tts_queue_to_elevenlabs(self):
        while True:
            chunk = await self.tts_queue.get()
            if self.is_interrupted:
                continue
            if self.elevenlabs_ws:
                try:
                    payload = {"text": chunk, "try_trigger_generation": True}
                    await self.elevenlabs_ws.send(json.dumps(payload))
                except websockets.exceptions.ConnectionClosed:
                    break

    async def receive_elevenlabs_audio(self):
        if not self.elevenlabs_ws:
            return
        try:
            async for message in self.elevenlabs_ws:
                if self.is_interrupted:
                    continue
                data = json.loads(message)
                if data.get("audio"):
                    # audio is already base64 string
                    audio_b64 = data["audio"]
                    # Send to twilio
                    twilio_msg = {
                        "event": "media",
                        "streamSid": self.stream_sid,
                        "media": {
                            "payload": audio_b64
                        }
                    }
                    await self.twilio_ws.send_text(json.dumps(twilio_msg))
        except websockets.exceptions.ConnectionClosed:
            print("Elevenlabs socket closed")

    async def close(self):
        if self.dg_connection:
            await self.dg_connection.finish()
        if self.elevenlabs_ws:
            await self.elevenlabs_ws.send(json.dumps({"text": ""})) # Close signal
            await self.elevenlabs_ws.close()

File: backend/src/voiceAgent.ts
import WebSocket from 'ws';
import { createClient, LiveClient, LiveTranscriptionEvents } from '@deepgram/sdk';
import { GoogleGenAI } from '@google/genai';

const DEEPGRAM_API_KEY = process.env.DEEPGRAM_API_KEY || '';
const GEMINI_API_KEY = process.env.GEMINI_API_KEY || '';
const ELEVENLABS_API_KEY = process.env.ELEVENLABS_API_KEY || '';
const ELEVENLABS_VOICE_ID = process.env.ELEVENLABS_VOICE_ID || '21m00Tcm4TlvDq8ikWAM'; // Rachel

let ai: GoogleGenAI | null = null;
if (GEMINI_API_KEY) {
  ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });
}

export class VoiceAgentConnection {
  private twilioWs: WebSocket;
  private streamSid: string;
  private context: string;
  private pitch: string;

  private dgClient: LiveClient | null = null;
  private elevenlabsWs: WebSocket | null = null;

  private transcriptBuffer: string = '';
  private isInterrupted: boolean = false;
  private ttsQueue: string[] = [];
  private ttsProcessing: boolean = false;
  
  private currentLlmAbortController: AbortController | null = null;

  private fullTranscript: string = '';

  constructor(twilioWs: WebSocket, streamSid: string, context: string, pitch: string, private contactId: number, private db: any) {
    this.twilioWs = twilioWs;
    this.streamSid = streamSid;
    this.context = context;
    this.pitch = pitch;
  }

  public async start() {
    this.startDeepgram();
    this.startElevenLabs();
    this.processTtsQueue();
  }

  private startDeepgram() {
    const deepgram = createClient(DEEPGRAM_API_KEY);
    this.dgClient = deepgram.listen.live({
      model: 'nova-2',
      language: 'en-US',
      encoding: 'mulaw',
      channels: 1,
      sample_rate: 8000,
      smart_format: true,
      filler_words: false,
      interim_results: false,
      utterance_end_ms: 1000,
      vad_events: true
    });

    this.dgClient.on(LiveTranscriptionEvents.Open, () => {
      console.log('Deepgram connection opened');
    });

    this.dgClient.on(LiveTranscriptionEvents.Transcript, (data: any) => {
      const sentence = data.channel.alternatives[0].transcript;
      if (sentence && data.is_final) {
        console.log(`User heard: ${sentence}`);
        this.transcriptBuffer += ' ' + sentence;
        this.fullTranscript += `\nUser: ${sentence}`;
        this.isInterrupted = false;

        this.generateResponse(this.transcriptBuffer);
        this.transcriptBuffer = '';
      }
    });

    this.dgClient.on(LiveTranscriptionEvents.SpeechStarted, () => {
      console.log('Barge-in detected!');
      this.isInterrupted = true;
      if (this.currentLlmAbortController) this.currentLlmAbortController.abort();
      this.twilioWs.send(JSON.stringify({ event: 'clear', streamSid: this.streamSid }));
      this.ttsQueue = [];
    });
  }

  private startElevenLabs() {
    const url = `wss://api.elevenlabs.io/v1/text-to-speech/${ELEVENLABS_VOICE_ID}/stream-input?model_id=eleven_monolingual_v1&output_format=ulaw_8000`;
    this.elevenlabsWs = new WebSocket(url);

    this.elevenlabsWs.on('open', () => {
      this.elevenlabsWs?.send(JSON.stringify({ text: " ", xi_api_key: ELEVENLABS_API_KEY }));
    });

    this.elevenlabsWs.on('message', (message) => {
      if (this.isInterrupted) return;
      const data = JSON.parse(message.toString());
      if (data.audio) {
        this.twilioWs.send(JSON.stringify({ event: "media", streamSid: this.streamSid, media: { payload: data.audio } }));
      }
    });
  }

  public processAudioFromTwilio(audioBytes: Buffer) {
    if (this.dgClient?.getReadyState() === 1) this.dgClient.send(audioBytes as any);
  }

  private async generateResponse(text: string) {
    if (!text.trim() || this.isInterrupted || !ai) return;
    if (this.currentLlmAbortController) this.currentLlmAbortController.abort();
    this.currentLlmAbortController = new AbortController();

    try {
      const prompt = `System: You are an AI sales agent. Objective: ${this.pitch}. Info: ${this.context}. Keep brief, natural, and stay on topic. User: ${text}`;
      const responseStream = await ai.models.generateContentStream({ model: 'gemini-1.5-flash', contents: prompt });

      let sentenceChunk = '';
      for await (const chunk of responseStream) {
        if (this.isInterrupted || this.currentLlmAbortController?.signal.aborted) break;
        if (chunk.text) {
          sentenceChunk += chunk.text;
          if (/[.?!\\n]/.test(sentenceChunk)) {
            this.ttsQueue.push(sentenceChunk);
            this.fullTranscript += `\nAI: ${sentenceChunk}`;
            sentenceChunk = '';
          }
        }
      }
      if (sentenceChunk && !this.isInterrupted) {
        this.ttsQueue.push(sentenceChunk);
        this.fullTranscript += `\nAI: ${sentenceChunk}`;
      }
    } catch (e: any) {
      if (e.name !== 'AbortError') console.error('LLM Error:', e);
    }
  }

  private async processTtsQueue() {
    setInterval(() => {
      if (this.isInterrupted) {
        this.ttsQueue = []; return;
      }
      if (this.ttsQueue.length > 0 && this.elevenlabsWs?.readyState === WebSocket.OPEN) {
        const chunk = this.ttsQueue.shift();
        if (chunk) this.elevenlabsWs.send(JSON.stringify({ text: chunk, try_trigger_generation: true }));
      }
    }, 50);
  }

  public async close() {
    if (this.dgClient) this.dgClient.finish();
    if (this.elevenlabsWs) {
      this.elevenlabsWs.send(JSON.stringify({ text: "" }));
      this.elevenlabsWs.close();
    }
    if (this.currentLlmAbortController) this.currentLlmAbortController.abort();
    
    // Save summarization
    if (this.fullTranscript.trim() && ai) {
      try {
        const summary = await ai.models.generateContent({
           model: 'gemini-1.5-flash',
           contents: `Summarize this call transcript into a few professional notes for a CRM. Indicate if the client was interested.\n\nTranscript:\n${this.fullTranscript}`
        });
        const note = summary.text || 'Call finished.';
        this.db.prepare(`UPDATE contacts SET notes = ? WHERE id = ?`).run(note, this.contactId);
      } catch (e) { console.error('Summary failed:', e); }
    }
  }
}
